File: androidApp/src/main/java/org/blackcandy/android/bridge/FlashComponent.kt
package org.blackcandy.android.bridge

import dev.hotwire.core.bridge.BridgeComponent
import dev.hotwire.core.bridge.BridgeDelegate
import dev.hotwire.core.bridge.Message
import dev.hotwire.navigation.destinations.HotwireDestination
import kotlinx.serialization.Serializable
import org.blackcandy.android.fragments.web.WebFragment
import org.blackcandy.shared.viewmodels.WebViewModel

class FlashComponent(
    name: String,
    private val delegate: BridgeDelegate<HotwireDestination>,
) : BridgeComponent<HotwireDestination>(name, delegate) {
    private val viewModel: WebViewModel
        get() {
            val fragment = delegate.destination.fragment as WebFragment
            return fragment.viewModel
        }

    override fun onReceive(message: Message) {
        when (message.event) {
            "connect" -> handleConnectEvent(message)
        }
    }

    private fun handleConnectEvent(message: Message) {
        val data = message.data<MessageData>() ?: return
        viewModel.showFlashMessage(data.message)
    }

    @Serializable
    data class MessageData(
        val message: String,
    )
}

File: iosApp/iosApp/Bridge/FlashComponent.swift
import HotwireNative
import sharedKit

class FlashComponent: BridgeComponent {
    override class var name: String { "flash" }

    private var viewModel: WebViewModel? {
        let viewController = delegate?.destination as? WebViewController
        return viewController?.viewModel
    }

    override func onReceive(message: Message) {
        switch message.event {
        case "connect":
            handleConnectEvent(message)
        default:
            break
        }
    }

    private func handleConnectEvent(_ message: Message) {
        guard let data: MessageData = message.data() else { return }
        viewModel?.showFlashMessage(message: data.message)
    }
}

extension FlashComponent {
    struct MessageData: Decodable {
        let message: String
    }
}
